File: python/numbers/binary.py
"""Binary to Decimal and Back — convert between number systems."""


def to_binary(n: int) -> str:
    """Convert a non-negative integer to its binary string representation."""
    if n < 0:
        raise ValueError("n must be non-negative")
    if n == 0:
        return "0"
    bits = []
    while n > 0:
        bits.append(str(n & 1))
        n >>= 1
    return "".join(reversed(bits))


def to_decimal(binary: str) -> int:
    """Convert a binary string to its decimal integer value."""
    if not binary:
        raise ValueError("empty string")
    if any(c not in "01" for c in binary):
        raise ValueError(f"invalid binary string: {binary}")
    result = 0
    for bit in binary:
        result = (result << 1) | int(bit)
    return result


def to_base(n: int, base: int) -> str:
    """Convert a non-negative integer to an arbitrary base (2-36)."""
    if base < 2 or base > 36:
        raise ValueError("base must be between 2 and 36")
    if n < 0:
        raise ValueError("n must be non-negative")
    if n == 0:
        return "0"
    digits = "0123456789abcdefghijklmnopqrstuvwxyz"
    result = []
    while n > 0:
        result.append(digits[n % base])
        n //= base
    return "".join(reversed(result))


def from_base(s: str, base: int) -> int:
    """Convert a string in an arbitrary base (2-36) to decimal."""
    if base < 2 or base > 36:
        raise ValueError("base must be between 2 and 36")
    if not s:
        raise ValueError("empty string")
    digits = "0123456789abcdefghijklmnopqrstuvwxyz"
    result = 0
    for c in s.lower():
        val = digits.index(c)
        if val >= base:
            raise ValueError(f"digit '{c}' invalid for base {base}")
        result = result * base + val
    return result


if __name__ == "__main__":
    print(f"42 -> binary: {to_binary(42)}")
    print(f"101010 -> decimal: {to_decimal('101010')}")
    print(f"255 -> hex: {to_base(255, 16)}")
    print(f"ff -> decimal: {from_base('ff', 16)}")

File: rust/numbers/binary.rs
//! Binary to Decimal and Back — convert between number systems.

/// Convert a non-negative integer to its binary string.
pub fn to_binary(n: u64) -> String {
    if n == 0 {
        return "0".to_string();
    }
    let mut bits = Vec::new();
    let mut val = n;
    while val > 0 {
        bits.push(if val & 1 == 1 { '1' } else { '0' });
        val >>= 1;
    }
    bits.into_iter().rev().collect()
}

/// Convert a binary string to its decimal value.
pub fn to_decimal(binary: &str) -> Result<u64, String> {
    if binary.is_empty() {
        return Err("empty string".to_string());
    }
    let mut result: u64 = 0;
    for c in binary.chars() {
        match c {
            '0' | '1' => result = (result << 1) | (c as u64 - '0' as u64),
            _ => return Err(format!("invalid binary digit: {c}")),
        }
    }
    Ok(result)
}

/// Convert a non-negative integer to a string in the given base (2-36).
pub fn to_base(n: u64, base: u64) -> Result<String, String> {
    if !(2..=36).contains(&base) {
        return Err("base must be between 2 and 36".to_string());
    }
    if n == 0 {
        return Ok("0".to_string());
    }
    let digits: Vec<char> = "0123456789abcdefghijklmnopqrstuvwxyz".chars().collect();
    let mut result = Vec::new();
    let mut val = n;
    while val > 0 {
        result.push(digits[(val % base) as usize]);
        val /= base;
    }
    Ok(result.into_iter().rev().collect())
}

/// Convert a string in the given base (2-36) to decimal.
pub fn from_base(s: &str, base: u64) -> Result<u64, String> {
    if !(2..=36).contains(&base) {
        return Err("base must be between 2 and 36".to_string());
    }
    if s.is_empty() {
        return Err("empty string".to_string());
    }
    let mut result: u64 = 0;
    for c in s.to_lowercase().chars() {
        let val = match c {
            '0'..='9' => c as u64 - '0' as u64,
            'a'..='z' => c as u64 - 'a' as u64 + 10,
            _ => return Err(format!("invalid digit: {c}")),
        };
        if val >= base {
            return Err(format!("digit '{c}' invalid for base {base}"));
        }
        result = result * base + val;
    }
    Ok(result)
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn roundtrip_binary() {
        assert_eq!(to_decimal(&to_binary(42)).unwrap(), 42);
    }

    #[test]
    fn roundtrip_hex() {
        assert_eq!(from_base(&to_base(255, 16).unwrap(), 16).unwrap(), 255);
    }

    #[test]
    fn zero() {
        assert_eq!(to_binary(0), "0");
        assert_eq!(to_decimal("0").unwrap(), 0);
    }
}
